EIPS/eip-7937.md

```diff
@@ -0,0 +1,145 @@
+
+* `JUMP` will only read the last 64 bits from the stack value. The rest 192 bits are discarded without reading.
+* `JUMPI` will only read the last 64 bits from the stack as destination, and the condition check will only read the last 64 bits.
+* The gas costs for `JUMP` and `JUMPI` remains the same.
+
+### EOFv1 mode and RJUMPI, RJUMPV
+
+If EOFv1 mode is entered, then an additional validation step is added. If the opcode `C0` is encountered and it is not part of PUSH opcode's data, then the interpreter MUST validate that:
+
+* The next opcode exists.
+* The next opcode is one of the valid 64-bit mode opcode described above.
+
+For flow operations RJUMPI and RJUMPV, the 64-bit mode has following changes:
+
+* For `RJUMPI`, the condition popped from stack is only read for the last 64 bits.
+* For `RJUMPV`, the case popped from stack is only read for the last 64 bits.
+
+Note that:
+
+* `RJUMP` is automatically in 64-bit mode because it does not read or write the stack.
+* The gas costs for `RJUMPI` and `RJUMPV` remains the same.
+
+## Rationale
+
+When a smart contract uses the 64-bit mode, it's expected that once entered, it will want to stay in 64-bit mode, and only exit to non-64-bit mode when the computationally intensive function is finished. This EIP is designed particularly with this fact in mind.
+
+All 64-bit opcodes only operates on the 64-bit value. It totally discards the rest 192 bits. The interpreter only needs to ensure that when it exits into non-64-bit mode and next time when a value result is read, that value has the first 192 bits reset to zero. The EVM interpreter can therefore use a typed stack for optmization:
+
+```haskell
+type StackItem = Value U256 | Value64 U64
+```
+
+The typed stack can also be implemented as a bitmap for memory alignment.
+
+For all inputs of 64-bit opcodes, it will either read in a `Value`, when it'll only take the last 64 bits, or a `Value64`, which is what is needed. It then always outputs a `Value64`. After exiting into non-64-bit mode and upon a `Value64` is read, the interpreter then translate the value back to 256-bit `Value` by extending the first 192 bits with zero.
+
+The 64-bit mode does not contain any opcodes which depends on the value's endianness, therefore `Value64` can also be stored in the optimal endianness of the architecture.
+
+The 64-bit mode will not save any memory usage.
+
+### Recommendations
+
+This EIP also recommends that we reserve `C0`-`CF` for prefix (modes) opcodes. For example, an additional modes OVERFLOW can be envisioned that changes the behavior of arithmetic opcodes from wrapping to overflow OOG, which can help to reduce, for example, the extra cycles needed for `SafeMath`.
+
+## Backwards Compatibility
+
+This EIP introduces a new (prefix) opcode `C0`. `C0` was previously an invalid opcode that has little usage, and thus the backward compatibility issues are minimal.
+
+## Test Cases
+
+To be added.
```

Comment:
Also as examples these would be helpful, I am somewhat confused by how certain operations would work if the original value has most bits of the 256 available set (instead of only the available 64). For instance operations like shift right (SAR).